lib/Sema/CSDiagnostics.cpp

```diff
@@ -1328,6 +1328,37 @@ bool MissingExplicitConversionFailure::diagnoseAsError() {
   return true;
 }
 
+/// Returns true if \p locator is ending with either of the following
+///  - Member
+///  - Member -> KeyPathDynamicMember
+static bool isLocatorPointingToMember(ConstraintLocator *locator) {
+  if (locator->isLastElement<LocatorPathElt::Member>()) {
+    return true;
+  } else if (locator->isLastElement<LocatorPathElt::KeyPathDynamicMember>()) {
+    auto path = locator->getPath();
+    if (path.size() >= 2 &&
+        path[path.size() - 2].is<LocatorPathElt::Member>()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/// Shared implementation of \c OnMemberBaseDiagnostic::getAnchor and
+/// \c PropertyWrapperReferenceFailure::getAnchor.
+static ASTNode getMemberBaseAnchor(ASTNode anchor, ConstraintLocator *locator) {
+  if (isLocatorPointingToMember(locator)) {
+    if (auto *UDE = getAsExpr<UnresolvedDotExpr>(anchor)) {
+      return UDE->getBase();
+    }
+  }
+  return anchor;
+}
+
+ASTNode OnMemberBaseDiagnostic::getAnchor() const {
+  return getMemberBaseAnchor(FailureDiagnostic::getAnchor(), getLocator());
```

Comment:
Anything that simplifies down to `UnresolvedDotExpr` should get its base expression returned right?

Reply:
Nope, in test/Constraints/diagnostics.swift:236 we have the following

```swift
func testStructWithOptionalArray(_ foo: StructWithOptionalArray) -> Int {
  return foo.array[0]  // expected-error {{value of optional type '[Int]?' must be unwrapped to refer to member 'subscript' of wrapped base type '[Int]'}}
  // expected-note@-1{{chain the optional using '?' to access member 'subscript' only for non-'nil' base values}}{{19-19=?}}
  // expected-note@-2{{force-unwrap using '!' to abort execution if the optional value contains 'nil'}}{{19-19=!}}
}
```

The locator for the fix is the following, which resolves to the base of the subscript (possibly another, similar problem).
```
locator@0x7fc4c780f428 [Subscript@/Users/alex/swift-src/swift/test/Constraints/diagnostics.swift:8:19 -> subscript member]
```

And because it’s the subscript member we want to insert the the `!` right before the subscript access and not before after `foo`.

Reply:
What is the `getAnchor()` going to be in this case? I think it's going to be `SubscriptExpr` which is okay because I'm only proposing calling `getBase()` if anchor is `UnresolvedDotExpr`?

Reply:
`getAnchor()` is the `UnresolvedDotExpr` that we are calling the subscript on.

Reply:
For pasterity - the was a check in FailureDiangostic::getAnchor which was causing this, the check is now removed.